File: python/two_sum_2.py
class Solution:
    def twoSum(self, numbers: list[int], target: int) -> list[int]:
        """time complexity: O(n), space complexity: O(1)"""
        l: int = 0
        r: int = len(numbers) - 1
        while l < r:
            if numbers[l] + numbers[r] < target:
                l += 1
            elif numbers[l] + numbers[r] > target:
                r -= 1
            else:
                return [l + 1, r + 1]

        return []


def test_two_sum_2_case1():
    # arrange
    numbers: list[int] = [2, 7, 11, 15]
    target: int = 9
    expected: list[int] = [1, 2]

    # act
    solution = Solution()
    actual = solution.twoSum(numbers, target)

    # assert
    actual.sort()
    expected.sort()
    assert actual == expected


def test_two_sum_2_case2():
    # arrange
    numbers: list[int] = [2, 3, 4]
    target: int = 6
    expected: list[int] = [1, 3]

    # act
    solution = Solution()
    actual = solution.twoSum(numbers, target)

    # assert
    actual.sort()
    expected.sort()
    assert actual == expected


def test_two_sum_2_case3():
    # arrange
    numbers: list[int] = [-1, 0]
    target: int = -1
    expected: list[int] = [1, 2]

    # act
    solution = Solution()
    actual = solution.twoSum(numbers, target)

    # assert
    actual.sort()
    expected.sort()
    assert actual == expected

File: rust/src/code/two_sum_2.rs
struct Solution {}

impl Solution {
    pub fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
        let mut l: usize = 0;
        let mut r: usize = numbers.len() - 1;
        while l < r {
            if numbers[l] + numbers[r] > target {
                r -= 1;
            } else if numbers[l] + numbers[r] < target {
                l += 1;
            } else if numbers[l] + numbers[r] == target {
                return Vec::from([l as i32 + 1, r as i32 + 1]);
            }
        }

        return Vec::new();
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn two_sum_2_case_1() {
        // arrange
        let numbers = vec![2, 7, 11, 15];
        let target = 9;
        let expected = vec![1, 2];

        // act
        let actual = Solution::two_sum(numbers, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn two_sum_2_case_2() {
        // arrange
        let numbers = vec![2, 3, 4];
        let target = 6;
        let expected = vec![1, 3];

        // act
        let actual = Solution::two_sum(numbers, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn two_sum_2_case_3() {
        // arrange
        let numbers = vec![-1, 0];
        let target = -1;
        let expected = vec![1, 2];

        // act
        let actual = Solution::two_sum(numbers, target);

        // assert
        assert_eq!(expected, actual);
    }
}
